scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/job/termination/handlers/TerminateReplicateTaskHandler.java

```diff
@@ -0,0 +1,240 @@
+					InternalTask nt = it.getValue();
+					nt.setJobInfo(((JobInfoImpl) internalJob.getJobInfo()));
+					int dupIndex = getNextReplicationIndex(InternalTask.getInitialName(nt.getName()),
+							nt.getIterationIndex());
+					internalJob.addTask(nt);
+					nt.setReplicationIndex(dupIndex);
+					assignReplicationTag(nt, initiator, false, action);
+				}
+				changesInfo.newTasksAdded(dup.values());
+
+				// find the beginning and the ending of the replicated block
+				for (Entry<TaskId, InternalTask> it : dup.entrySet()) {
+					InternalTask nt = it.getValue();
+
+					// connect the first task of the replicated block to the
+					// initiator
+					if (todup.getId().equals(it.getKey())) {
+						newTarget = nt;
+						newTarget.addDependence(initiator);
+						// no need to add newTarget to modifiedTasks
+						// because newTarget is among dup.values(), and we
+						// have added them all
+					}
+					// connect the last task of the block with the merge task(s)
+					if (target.getId().equals(it.getKey())) {
+						newEnd = nt;
+
+						List<InternalTask> toAdd = new ArrayList<>();
+						// find the merge tasks ; can be multiple
+						for (InternalTask t : internalJob.getIHMTasks().values()) {
+							List<InternalTask> pdeps = t.getIDependences();
+							if (pdeps != null) {
+								for (InternalTask parent : pdeps) {
+									if (parent.getId().equals(target.getId())) {
+										toAdd.add(t);
+									}
+								}
+							}
+						}
+						// connect the merge tasks
+						for (InternalTask t : toAdd) {
+							t.addDependence(newEnd);
+							changesInfo.taskUpdated(t);
+						}
+					}
+				}
+
+				// propagate the changes on the JobDescriptor
+				internalJob.getJobDescriptor().doReplicate(taskId, dup, newTarget, target.getId(), newEnd.getId());
+
+			}
+		}
+
+		// notify frontend that tasks were added to the job
+		((JobInfoImpl) internalJob.getJobInfo()).setTasksChanges(changesInfo, internalJob);
+		if (frontend != null) {
+			frontend.jobStateUpdated(internalJob.getOwner(),
+					new NotificationData<>(SchedulerEvent.TASK_REPLICATED, internalJob.getJobInfo()));
+		}
+		((JobInfoImpl) internalJob.getJobInfo()).clearTasksChanges();
+
+		// no jump is performed ; now that the tasks have been replicated and
+		// configured, the flow can continue its normal operation
+		internalJob.getJobDescriptor().terminate(taskId);
+
+		return true;
+	}
+
+	/**
+	 * Assign a tag to new duplicated task because of a REPLICATE or LOOP.
+	 * 
+	 * @param replicatedTask
+	 *            the new duplicated task.
+	 * @param initiator
+	 *            the initiator of the duplication.
+	 * @param loopAction
+	 *            true if the duplication if after a loop or, false if it is a
+	 *            replicate.
+	 * @param action
+	 *            the duplication action.
+	 */
+	private void assignReplicationTag(InternalTask replicatedTask, InternalTask initiator, boolean loopAction,
+			FlowAction action) {
+		StringBuffer buf = new StringBuffer();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace the synchronized class "StringBuffer" by an unsynchronized one such as "StringBuilder". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1149)